src/uint/bingcd/gcd.rs

```diff
@@ -0,0 +1,255 @@
+use crate::{NonZero, Odd, Uint, U128, U64};
+
+impl<const LIMBS: usize> NonZero<Uint<LIMBS>> {
+    /// Compute the greatest common divisor of `self` and `rhs`.
+    pub const fn bingcd(&self, rhs: &Uint<LIMBS>) -> Self {
+        let val = self.as_ref();
+        // Leverage two GCD identity rules to make self and rhs odd.
+        // 1) gcd(2a, 2b) = 2 * gcd(a, b)
+        // 2) gcd(a, 2b) = gcd(a, b) if a is odd.
+        let i = val.is_nonzero().select_u32(0, val.trailing_zeros());
+        let j = rhs.is_nonzero().select_u32(0, rhs.trailing_zeros());
+        let k = const_min(i, j);
+
+        val.shr(i)
+            .to_odd()
+            .expect("val.shr(i) is odd by construction")
+            .bingcd(rhs)
+            .as_ref()
+            .shl(k)
+            .to_nz()
+            .expect("gcd of non-zero element with another element is non-zero")
+    }
+}
+
+impl<const LIMBS: usize> Odd<Uint<LIMBS>> {
+    const BITS: u32 = Uint::<LIMBS>::BITS;
+
+    /// Compute the greatest common divisor of `self` and `rhs` using the Binary GCD algorithm.
+    ///
+    /// This function switches between the "classic" and "optimized" algorithm at a best-effort
+    /// threshold. When using [Uint]s with `LIMBS` close to the threshold, it may be useful to
+    /// manually test whether the classic or optimized algorithm is faster for your machine.
+    #[inline(always)]
+    pub const fn bingcd(&self, rhs: &Uint<LIMBS>) -> Self {
+        // Todo: tweak this threshold
+        // Note: we're swapping the parameters here for greater readability: Pornin's Algorithm 1
+        // and Algorithm 2 both require the second argument (m) to be odd. Given that the gcd
+        // is the same, regardless of the order of the parameters, this swap does not affect the
+        // result.
+        if LIMBS < 8 {
+            self.classic_bingcd(rhs)
+        } else {
+            self.optimized_bingcd(rhs)
+        }
+    }
+
+    /// Computes `gcd(self, rhs)`, leveraging the (a constant time implementation of) the classic
+    /// Binary GCD algorithm.
+    ///
+    /// Note: this algorithm is efficient for [Uint]s with relatively few `LIMBS`.
+    ///
+    /// Ref: Pornin, Optimized Binary GCD for Modular Inversion, Algorithm 1.
+    /// <https://eprint.iacr.org/2020/972.pdf>
+    #[inline]
+    pub const fn classic_bingcd(&self, rhs: &Uint<LIMBS>) -> Self {
+        // (self, rhs) corresponds to (m, y) in the Algorithm 1 notation.
+        let (mut a, mut b) = (*rhs, *self.as_ref());
+        let mut j = 0;
+        while j < (2 * Self::BITS - 1) {
+            j += 1;
+
+            let a_odd = a.is_odd();
+
+            // swap if a odd and a < b
+            let a_lt_b = Uint::lt(&a, &b);
+            let do_swap = a_odd.and(a_lt_b);
+            Uint::conditional_swap(&mut a, &mut b, do_swap);
+
+            // subtract b from a when a is odd
+            a = a.wrapping_sub(&Uint::select(&Uint::ZERO, &b, a_odd));
```

Comment:
@tarcieri what do you think of this compared to how I had it previously:

```suggestion
            a = Uint::select(&a, &a.wrapping_sub(&b), a_odd);
```

it yields 25-10% improvement for Uints with few limbs

Reply:
I'm surprised there's that much of a difference. Are you sure it's always faster or is it faster depending on `a`?

Reply:
I'm pretty confident it is always faster.

The reason I think this is faster, is that we are now selecting between a constant and a variable, instead of two variables. Given that `select` is `const` and loves to be inlined, the compiler can now optimize the `select` operation. 

Recall, `Uint::select` calls `Limb::select`, which in turn calls

```rust
impl ConstChoice {
    /// Return `b` if `self` is truthy, otherwise return `a`.
    #[inline]
    pub(crate) const fn select_word(&self, a: Word, b: Word) -> Word {
        a ^ (self.0 & (a ^ b))
    }
}
```

When `a` is the constant ZERO, this can be optimized as:

```rust
        self.0 & b
```

saving two XOR operations, or 2/3's of this operation.

Returning to the `gcd` subroutine, this `select` is in the hot loop of this algorithm. In total, the loop executes:
- `Uint::is_odd` (1 op)
- `Uint::lt` (2 ops/word), 
- `ConstChoice::and` (1 op), 
- `Uint::wrapping_sub` (2 ops/word),
- `Uint::select` (3 ops/word -> 1 ops/word)
- `Uint::shr` (3 ops/word)

So, there is a reduction from 10 to 8 ops/word, or a 20% improvement.

Reply:
> When a is the constant ZERO

Aah, ok